python/GafferSceneUI/SetFilterUI.py

```diff
@@ -94,3 +98,148 @@
 	}
 
 )
+
+##########################################################################
+# NodeGadget drop handler
+##########################################################################
+
+__DropMode = enum.Enum( "__DropMode", [ "None_", "Add", "Remove", "Replace" ] )
+
+__originalDragPointer = None
+
+def __setsPlug( node ) :
+
+	for plug in node.children( Gaffer.Plug ) :
+		if Gaffer.Metadata.value( plug, "ui:scene:acceptsSetNames" ) or Gaffer.Metadata.value( plug, "ui:scene:acceptsSetExpression" ) :
+			return plug
+
+	return None
+
+def __dropMode( nodeGadget, event ) :
+
+	if __setsPlug( nodeGadget.node() ) is None :
+		filter = None
+		if nodeGadget.node()["filter"].getInput() is not None :
+			filter = nodeGadget.node()["filter"].source().node()
+		if filter is None :
+			return __DropMode.Replace
+		elif not isinstance( filter, GafferScene.SetFilter ) :
+			return __DropMode.None_
+
+	if event.modifiers & event.Modifiers.Shift :
+		return __DropMode.Add
+	elif event.modifiers & event.Modifiers.Control :
+		return __DropMode.Remove
+	else :
+		return __DropMode.Replace
+
+def __expressionSets( setExpression ) :
+
+	if (
+		"in" in setExpression or
+		"containing" in setExpression or
+		"|" in setExpression or
+		"&" in setExpression or
+		"(" in setExpression
+	) :
+		return [ f"( {setExpression} )" ]
+
+	return setExpression.split( " " )
+
+def __dragEnter( nodeGadget, event ) :
+
+	if not isinstance( event.data, IECore.StringVectorData ) :
+		return False
+
+	if not len ( event.data ) :
+		return False
+
+	if __dropMode( nodeGadget, event ) == __DropMode.None_ :
+		return False
+
+	global __originalDragPointer
+	__originalDragPointer = GafferUI.Pointer.getCurrent()
+
+	return True
+
+def __dragLeave( nodeGadget, event ) :
+
+	global __originalDragPointer
+
+	GafferUI.Pointer.setCurrent( __originalDragPointer )
+	__originalDragPointer = None
+
+	return True
+
+def __dragMove( nodeGadget, event ) :
+
+	global __originalDragPointer
+	if __originalDragPointer is None :
+		return False
+
+	GafferUI.Pointer.setCurrent( __dropMode( nodeGadget, event ).name.lower() + "Paths" )
+
+	return True
+
+def __drop( nodeGadget, event ) :
+
+	global __originalDragPointer
+	if __originalDragPointer is None :
+		return False
+
+	setsPlug = __setsPlug( nodeGadget.node() )
+	if setsPlug is None :
+		setsPlug = __setsPlug( nodeGadget.node()["filter"].source().node() )
+
+	dropSets = event.data
+
+	dropMode = __dropMode( nodeGadget, event )
+	if dropMode == __DropMode.Replace :
+		sets = sorted( dropSets )
+	elif dropMode == __DropMode.Add :
+		sets = set( __expressionSets( setsPlug.getValue() ) )
+		sets.update( dropSets )
+		sets = sorted( sets )
+	else :
+		sets = set( __expressionSets( setsPlug.getValue() ) )
+		for s in dropSets :
+			if s in sets :
+				sets.remove( s )
+			else :
+				joinedExpression = " ".join( sets )
```

Comment:
We're losing the original ordering here.

Reply:
We were also losing the order when adding sets. I've maintained the order in both in https://github.com/GafferHQ/gaffer/pull/5786/commits/ecdbab3a990d6082b50440856092d8cf2e7607b3

Reply:
I didn't explain this very well, but I quite liked the sorted order we were using when adding, and was hoping to do the same when removing. That way they're both consistent with what we're doing when Replacing. Whatever we do, I think all three operations should probably have the same approach to ordering.